src/models/gamma_poisson_model.py

```diff
@@ -247,116 +249,317 @@ def _impression_count(self, h):
+        # Append kplus_reach to h.  So, the last value of h is treated
+        # specially.
+        h.append(kplus_reach)
+        h = np.array(h)
+        return h
+
+    def _chi2_distance_with_one_plus_reach(
+        self, h, I, N, alpha, beta, one_plus_reach_weight=1.0
+    ):
+        """Returns distance between actual and expected histograms.
+
+        Computes the metric
+
+            chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i] +
+               w * (\sum_i (h[i] - hbar[i]))^2
+
+        where
+            hbar is the histogram that would be expected given parameters
+               I, N, alpha, beta
+            w is the one_plus_reach_weight that is used to determine how
+               much to weight an exact match of 1+ reach.
+
+        The first part of this objective formula is the chi-squared statistic
+        used for determining whether two sets of categorical samples are drawn
+        from the same underlying distribution.  Therefore, if the value returned
+        by this function is relatively low (say < 20), then there is reason to
+        believe that the Gamma-Poisson distribution adequately fits the data.
+
+        Args:
+          h: Actual frequency histogram.  h[i] is the number of people
+             reached exactly i+1 times, except for h[-1] which is the
+             number of people reached len(h) or more times.
+          I: Total number of impressions purchased.
+          N: Total number of unique users.
+          alpha:  Parameter of the Gamma-Poisson distribution.
+          beta:  Second parameter of the Gamma-Poisson distribution.
+          one_plus_reach_weight:  Weight to be given to the squared
+            difference in 1+ reach values for the actual and estimated
+            histograms.
+        Returns:
+          chi^2 distance between the actual and expected histograms, plus
+          a term for weighting the difference in 1+ reach.
+        """
+        # Estimate total number of potential impressions
+        Imax = self._expected_impressions(N, alpha, beta)
+        if Imax <= I:
+            return np.sum(np.array(h) ** 2)
+
+        hbar = list(self._expected_histogram(I, Imax, N, alpha, beta, len(h) - 1))
 
-    def _fit_histogram_chi2_distance(self, h):
+        # Compute expected number of people in the k+ bucket.
+        kplus_prob = 1.0 - np.sum(
+            self._kreach(np.arange(0, len(h)), I, Imax, alpha, beta)
+        )
+        hbar.append(kplus_prob * N)
+        hbar = np.array(hbar)
+
+        actual_oneplus = np.sum(h)
+        predicted_oneplus = np.sum(hbar)
+        oneplus_error = (actual_oneplus - predicted_oneplus) ** 2
+
+        obj = np.sum((hbar - h) ** 2 / (hbar + 1e-6)) + oneplus_error
+
+        if np.isnan(obj):
+            logging.vlog(2, f"alpha {alpha} beta {beta} N {N} Imax {Imax}")
+            logging.vlog(2, f"h    {h}")
+            logging.vlog(2, f"hbar {hbar}")
+            raise RuntimeError("Invalid value of objective function")
```

Comment:
The list of built-in Python exceptions can be found here:  https://docs.python.org/3/library/exceptions.html

In theory, the objective function should never produce NaN values.  If it does, then this indicates some sort of algorithmic problem.  I like RuntimeError because it is a generic error that captures this circumstance.

I suppose some of the other candidates that seem like they could be close matches are as follows:  FloatingPointError, OverflowError, ArithmeticError.  We don't really know what the cause of the error is, so FloatingPointError and OverflowError seem like they might not be accurate.  ArithmeticError seems like a reasonable choice, though.  

Let me know if you would like me to change the exception type.

Reply:
Thanks for the illustration!   Now I also feel that RuntimeError is appropriate.